Allow Latex expressions in single-line titles and Cartesian axes labels

PaulWessel opened the pull request:
**Description of proposed changes**

See #4563 for background.  This PR implements the ability to place _Latex_ expressions in titles and labels. Users who wish to use this feature must have a working installation of **latex** and **dvips** as well as recommended fonts.  Because a title may contain a mix of text and equation, I change the default _Latex_ font to match the corresponding title or label font chosen in GMT.  The correspondence should work for all of our standard fonts, e.g., Helvetica, Times-Roman, Courier, Bookman, Palatino, etc.

Some limitations that may be relaxed later:

1. Only single-line titles can have _Latex_ expressions in them.
2. The **subplot -T**_heading_ has not been updated for latex (nor multi-line headings) yet.
3. **colorbar** may also need some more work.

I have updated the documentation and added two test scripts (_latextitle.sh_ and _latexlabel.sh_).  The output of the latter looks like this:

![latexlabel](https://user-images.githubusercontent.com/26473567/102004730-5785cb80-3cb7-11eb-8803-a8965fc2e25f.png)

Feel free to test-drive this PR to possibly expose any issues.  Once acceptable we also need to update the run-time dependency entries on the wiki and Building.

PaulWessel commented:
I believe I have completed this pull request by also allowing Latex-expressions in pstext (except for paragraph mode **-M**).  I have added a new test latext.sh to demonstrate this latest feature:

![latext](https://user-images.githubusercontent.com/26473567/102685967-750bd700-4188-11eb-8572-1f09b592f8f2.png)

Please give this PR a review now, @joa-quim, @seisman and @meghanrjones.

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -320,7 +320,8 @@ Append **+n** to have no frame and annotations at all [Default is controlled by
 will appear centered above the plot frame; optionally also add a *subtitle* via **+s**.
 **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-using the markers '@^' or '#break;'.
+using the markers '@^' or '#break;'.  To include Latex code as part of a single-line title,
```
```suggestion
using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title,
```

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -320,7 +320,8 @@ Append **+n** to have no frame and annotations at all [Default is controlled by
 **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-using the markers '@^' or '#break;'.
+using the markers '@^' or '#break;'.  To include Latex code as part of a single-line title,
+enclose the expression with @$ markers (**Note**: Requires latex and dvips to be installed).
```
```suggestion
enclose the expression with @$ markers (**Note**: Requires ``latex`` and ``dvips`` to be installed).
```

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -345,6 +346,8 @@ but you may also split this into two separate invocations for clarity, i.e.,
     force a horizontal label for *y*-axes (useful for very short labels).
     For Cartesian axes you may specify an alternate via **+s** which is used for
     right or upper axis axis label (with any **+l** label used for left and bottom axes).
+    To include Latex code as part of a label, enclose the expression with @$ markers
```
```suggestion
    To include LaTeX code as part of a label, enclose the expression with @$ markers
```

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -345,6 +346,8 @@ but you may also split this into two separate invocations for clarity, i.e.,
     For Cartesian axes you may specify an alternate via **+s** which is used for
     right or upper axis axis label (with any **+l** label used for left and bottom axes).
+    To include Latex code as part of a label, enclose the expression with @$ markers
+    (**Note**: Requires latex and dvips to be installed).
```
```suggestion
    (**Note**: Requires ``latex`` and ``dvips`` to be installed).
```

seisman commented on doc/rst/source/explain_-B_full.rst_:
```diff
@@ -30,8 +30,9 @@
     **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-    using the markers '@^' or '#break;'.  The Frame setting is optional but
-    can be invoked once to override the above defaults.
+    using the markers '@^' or '#break;'.  To include Latex code as part of a single-line title,
```
```suggestion
    using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title,
```

seisman commented on doc/rst/source/explain_-B_full.rst_:
```diff
@@ -30,8 +30,9 @@
-    using the markers '@^' or '#break;'.  The Frame setting is optional but
-    can be invoked once to override the above defaults.
+    using the markers '@^' or '#break;'.  To include Latex code as part of a single-line title,
+    enclose the expression with @$ markers (**Note**: Requires latex and dvips to be installed).
```
```suggestion
    enclose the expression with @$ markers (**Note**: Requires ``latex`` and ``dvips``
     to be installed).
```

seisman commented on doc/rst/source/pstext.rst:
```diff
@@ -79,12 +73,15 @@ To add a typeset figure caption for a 3-inch wide illustration, use
+
+To place a line containing a Latex equation, try::
+
+    echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt pstext -R0/6/0/6 -JX15c -B -F+f32p+a30 -P > map.ps
```
In classic mode, we need **-Baf**, otherwise GMT reports `pstext [ERROR]: Found no history for option -`
```suggestion
    echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt pstext -R0/6/0/6 -JX15c -Baf -F+f32p+a30 -P > map.ps
```

seisman replied:
```suggestion
    echo 3 3 'Use @[\Delta g = 2\pi\rho Gh@[' | gmt pstext -R0/6/0/6 -JX15c -B -F+f32p+a30 -P > map.ps
```

seisman commented:
I don't have latex and dvips installed yet, and running the below command gives me:
```
$ echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt text -R0/6/0/6 -JX15c -B -F+f32p+a30 -pdf map 
text [ERROR]: latex is not installed or not in your executable path - cannot process Latex to DVI.
ho Gh@$" to EPS failedtext: Conversion of Latex "Use @$\Delta g = 2\pi
```
The error messages look a little messy.

seisman commented:
Please also do batch replacements from "Latex" to "LaTeX".

PaulWessel commented:
The product is called LaTeX but the executable is latex, so not sure about batch replace - think I have to do it per instance.

PaulWessel commented:
Hm, I get

```
echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt text -R0/6/0/6 -JX15c -B -F+f32p+a30 -pdf map 
text [ERROR]: dvips is not installed or not in your executable path - cannot convert DVI to EPS.
text [ERROR]: gmt_map_text: Conversion of LaTeX "Use @$\Delta g = 2\pi\rho Gh@$" to EPS failed
```

But I suppose I don't have to print out the text that failed - just the message?

seisman commented:
> Hm, I get
> 
> ```
> echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt text -R0/6/0/6 -JX15c -B -F+f32p+a30 -pdf map 
> text [ERROR]: dvips is not installed or not in your executable path - cannot convert DVI to EPS.
> text [ERROR]: gmt_map_text: Conversion of LaTeX "Use @$\Delta g = 2\pi\rho Gh@$" to EPS failed
> ```
> 
> But I suppose I don't have to print out the text that failed - just the message?

Perhaps because I'm using zsh and `\r` is interpreted by zsh shell. Yes, the message seems useless.

seisman commented:
Now I have latex and dvips installed, running the above example still doesn't work. Here are related debug message:
```
text [INFORMATION]: Reading Data Table from Standard Input stream
text [DEBUG]: GMT_Begin_IO: Input resource access is now enabled [record-by-record]
text [DEBUG]: Source col types: (Number,Number,String)
text [DEBUG]: ASCII source scanned: Numerical columns: 2, Trailing text: Y, Record type: Numerical with trailing text
text [DEBUG]: gmt_check_executable: Pass to popen: [latex --version 2> /dev/null]
text [DEBUG]: latex --version 2> /dev/null was successful
text [DEBUG]: latex found.
text [DEBUG]: gmt_check_executable: Pass to popen: [dvips --version 2> /dev/null]
text [DEBUG]: dvips --version 2> /dev/null was successful
text [DEBUG]: dvips found.
text [DEBUG]: gmtplot_latex_eps: Selecting font helvet [phv].
dvips: DVI file can't be opened: gmt_eq.dvi: No such file or directory
PSL:../../src/postscriptlight.c:6010: Error: Cannot open image file equation.eps!
text [ERROR]: gmtplot_latex_eps: Unable to load EPS file equation.eps
text [ERROR]: gmt_map_text: Conversion of LaTeX to EPS failed
text [DEBUG]: GMT_End_IO: Input resource access is now disabled
text [DEBUG]: Current size of half-baked PS file /Users/seisman/.gmt/sessions/gmt_session.28773/gmt_0.ps- = 24947.
text [INFORMATION]: pstext: Plotted 1 text strings
text [DEBUG]: GMT now running in modern mode [Session ID = 28773]
text [DEBUG]: Revised options: show
```

PaulWessel commented:
Hard for me to check if your latex distro has the recommended-fonts installed or not.  I wish I wrote down which ones I installed (latex-extra, latex-recommended-fonts?)

PaulWessel commented:
Looks like I installed (via macport):

```
texlive-basic 
texlive-bin-extra
texlive-fonts-recommended
texlive-latex 
texlive-latex-extra
texlive-latex-recommended

```

seisman commented:
I'm using [TinyTeX](https://yihui.org/tinytex/) to install a minimal TeXLive distribution, so I have many fonts missing now. I need to find the exact latex package names I need later. Before that, I think it would be good if we can improve the error messages when fonts are not available.

I tried to build this latex example using `latex eq.tex`:
```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{helvet}
\begin{document}
\thispagestyle{empty}
\fontfamily{phv}\selectfont
Solution to $\nabla^4 \Psi = \frac{\delta w}{\delta t} = f(x(t))$
\end{document}
```
It gives me the following error and **hangs up** (perhaps we should add `-interaction=nonstopmode` to the latex command?):
```
! LaTeX Error: File `helvet.sty' not found.

Type X to quit or <RETURN> to proceed,
or enter new name. (Default extension: sty)
```

joa-quim commented:
LaTex is not big. It is insanely big. 7.2 GB
![image](https://user-images.githubusercontent.com/537321/102724647-e5416680-4308-11eb-8deb-98ea504f8fcb.png)

Installed a minimal LaTex but that hangs the tests. Note that also that the tests are using single quotes instead of double. *E.g.* 
> -Byaf+l'@$\Delta\sigma_{xx}^2@$ (MPa)'


```
C:\v\build\test\psbasemap\latexlabel
$ gmt psbasemap -R-30/30/-20/20 -JX6i/2.5i -Byaf+l"@$\Delta\sigma_{xx}^2@$ (MPa)" -Bxaf+l"Evaluating @$\nabla^4 \Psi = \frac{\delta w}{\delta t} = f(x(t))@$" -B+t"The Main Results of my Research" -P -K -Xc -Y1.25i > lixo.ps
PSL:C:\progs_cygw\GMTdev\gmt5\master\src\postscriptlight.c:6010: Error: Cannot open image file equation.eps!
psbasemap [ERROR]: gmtplot_latex_eps: Unable to load EPS file equation.eps
psbasemap [ERROR]: gmt_map_label: Conversion of LaTeX label to EPS failed

kpathsea: Running mktextfm phvr8t

The command name is C:\programs\texlive\bin\win32\mktextfm
PSL:C:\progs_cygw\GMTdev\gmt5\master\src\postscriptlight.c:6010: Error: Cannot open image file equation.eps!
psbasemap [ERROR]: gmtplot_latex_eps: Unable to load EPS file equation.eps
psbasemap [ERROR]: gmt_map_label: Conversion of LaTeX label to EPS failed

kpathsea: Running mktexmf phvr8t.mf

The command name is C:\programs\texlive\bin\win32\mktexmf
name = phvr8t, rootname = phvr8t, pointsize =
mktexmf: empty or non-existent rootfile!
Cannot find phvr8t.mf.
kpathsea: Appending font creation commands to missfont.log.
...
```

seisman commented:
> LaTex is not big. It is insanely big. 7.2 GB

That's why I use the TinyTeX, only ~200 MB on macOS, but I have to install packages manually when I need.

> Note that also that the tests are using single quotes instead of double.

It's because the `$` sign is special for bash. It seems we have to use single quotes for bash and double quotes for batch. Complicated.

PaulWessel commented:
Yes, I will improve the error handling by adding the suggested -interaction=nonstopmode and if I get a system call error I will not delete the equation.tex file but instead give a message with its path and people can use it to test what to install.  We of course need to indicate on the wiki what the run-time requirements are if you want to use LaTeX.  Just got back from beach so running behind.

PaulWessel commented:
The `@$` is a problems due to shells.  Suggestions for a better system? Maybe `<math> ...</math`> or is there an unused `@<char`> option that is not special in batch or bash?

joa-quim commented:
I have 390 MB of Texish and still not enough. What does this mean? That I don't have that ``phvr8t.mf`` font?
```
C:\TMP\gmt_latex_d27260>latex gmt_eq.tex  1>NUL

kpathsea: Running mktextfm phvr8t

The command name is C:\programs\texlive\bin\win32\mktextfm

kpathsea: Running mktexmf phvr8t.mf

The command name is C:\programs\texlive\bin\win32\mktexmf
name = phvr8t, rootname = phvr8t, pointsize =
mktexmf: empty or non-existent rootfile!
Cannot find phvr8t.mf.
kpathsea: Appending font creation commands to missfont.log.
```

seisman commented:
> The `@$` is a problems due to shells. Suggestions for a better system? Maybe `<math> ...</math`> or is there an unused `@<char`> option that is not special in batch or bash?

https://en.wikibooks.org/wiki/LaTeX/Mathematics

![image](https://user-images.githubusercontent.com/3974108/102726012-275ab000-42e9-11eb-825e-c236c66d16e9.png)

Is `@(` and `@)` possible?

PaulWessel commented:
Parenthesis is a shell thing too, no?

joa-quim commented:
``<math> </math>`` is clean stuff (still programmer lingo but a short one)

PaulWessel commented:
```
echo @[stuff@] 
@[stuff@]
echo @(stuff@)
-bash: syntax error near unexpected token `('
```

Maybe @[ ... @[ (or close with @] could work?

PaulWessel commented:
For titles/subtitle we allowed @^ or` #break;`  So we should be consistent, either @[ and `#math; `or replace both with `<break> ...</break>` and `<math> ... </math>`

seisman commented:
`@(` works if enclosing with double quotes
```
$ echo "@(stuff@)"
@(stuff@)
```

joa-quim commented:
The ``<blabla> ...</blabla>``  is a much more known construct. Consistency should not drive us into cryptic escape sequences that tell nothing to users. Just that its the rule to do that thing.

PaulWessel commented:
Offering both `<math> `and `@[ `does not hurt anyone.  And you are all about brevity - witness your commands where you leave off the leading "gmt" before the module!

joa-quim commented:
I often leave the ``gmt`` because *I don't need it* and my brain rebels against doing non-needed things. But I', also the one that is pushing for the understandable long options. Sure we can offer both ``@(?)`` a clear option. I just think the second is enough.

PaulWessel commented:
The latest commit will inform the user when failure and state they should investigate why the script fails, and points out the temp directory with the script and results instead of deleting it.  E.g., 

```
ext [ERROR]: gmtplot_latex_eps: Running "sh gmt_eq.sh" returned error 0.
text [ERROR]: gmtplot_latex_eps: Please run it manually to learn what LaTeX packages you are missing.
text [ERROR]: gmtplot_latex_eps: The script and logs can be found in directory /tmp/gmt_latex_cWrI1v
text [ERROR]: gmt_map_text: Conversion of LaTeX to EPS failed
```

Is this good enough?  Looking at the :TeX code I write, the packages that users must install are listed here:

`\usepackage[T1]{fontenc}\usepackage[utf8]{inputenc}\usepackage{helvet}`

So fontenc, inputenc, plus the various fonts themselves (helvet, for instance).

PaulWessel commented:
Let me get your votes for` #math; vs <math> ...</math>` and then the same applies to our `break` call.  @joa-quim, I assume `@[` has no special DOS meaning, right?

joa-quim commented:
I guess not
```
C:\v>@[
'[' is not recognized as an internal or external command,
operable program or batch file.
```

PaulWessel commented:
Is it a problem to generate [, { etc on European keyboards? I know for Norwegian those keys have special letters like æ, ø, å.

PaulWessel commented:
For the Latex to match the various GMT standard fonts, the \usepackage{font} needs to be able to load these fonts:

`helvet, mathptmx, courier, symbol, avantgar, bookman, newcent, mathpazo, zapfchan, zapfding`

So I can add some more docs on this somewhere.  Maybe in pstext man page?

seisman commented:
> So I can add some more docs on this somewhere. Maybe in pstext man page?

I think the latex integration deserves a small section in the cookbook (maybe appendix), and link to that section in the pstext manpage.

PaulWessel commented:
OK, will do.  Are you OK changing` #break; to <break> .. </break>` to be consistent?  Working on this right now so need to know before I recode..

joa-quim commented:
Sure, ``<break>`` has always been my choice. But we don't need a ``</break>`` right?

PaulWessel commented:
Right, no need for </break>, must for the math.

joa-quim commented:
> Is it a problem to generate [, { etc on European keyboards? 

Should not. Those are fundamental key strokes for programming.

meghanrjones commented on doc/rst/source/text_common.rst_:
```diff
@@ -23,7 +23,8 @@ Appendix :ref:`Chart-Octal-Codes-for-Chars` in the GMT Technical Reference and C
 (does not work for strings with sub/super scripts, symbols, or composite
-characters, except in paragraph mode (**-M**)).
+characters, except in paragraph mode (**-M**)).  Finally, you may typeset
+Latex expressions provided they are enclosed in @$...@$.
```
```suggestion
Latex expressions provided they are enclosed in @$...@$ and both ``latex`` 
and ``dvips`` are installed.
```

PaulWessel replied:
I've already rewritten that section to mention @[ and <math> so I think I cannot merge this now since it differs from what is in the branch.

joa-quim commented:
Do we really need this line?
> \fontfamily{phv}\selectfont

Because without it the ``gmt_eq.bat`` runs fine.

PaulWessel commented:
I cannot tell you for sure, but I think for at least some of them it is required.  There is the font _name_ and the font _code_; the _phv_ is the code.  I followed what I learned from the web about this and it worked - I am not sure it is optional.  Remember, this is **not** about the equation but _regular_ text outside the equation.  Otherwise, it will all be set in the same font regardless of FONT_TITLE, etc.

PaulWessel commented:
I have added a new appendix and reference it from -B and pstext.  However, I get this message

`/Users/pwessel/GMTdev/gmt-dev/dbuild/doc/rst/source/cookbook/gmt-latex.rst: WARNING: document isn't included in any toctree`

even though as far as I can tell I treat gmt-latex.rst  the same as, e.g., postscript-fonts.rst.  The appendix looks like this:

![Untitled](https://user-images.githubusercontent.com/26473567/102729612-7991d600-42d5-11eb-898a-77f1f17db381.png)

PaulWessel commented:
On my virgin Mac mini M1 I with MacPorts I added these TexLive packages to get things to work

1. texlive-basic.  This installed many things, including the required dvips command
2. texlive-latex: Same, including the required latex command.
3. texlive-fonts-recommended: This gave me the fonts needed.

Now I could run 

`echo 3 3 "My text is @[\nabla^4 w = \Delta \sigma_{xx}^2@[" | gmt text -R0/6/0/6 -JX6i -B -pdf map`

successfully.  Note if the fonts are missing then the "My text is " part will **not** be set!

So listing those three packages for MacPorts on the wiki seems sensible.

joa-quim commented:
Bad things happen when one opens these latex PS

PaulWessel commented:
Well, they pass all tests for me so you need to be more specific.  Maybe post the PS for the command above so we can compare?

joa-quim commented:
It's not that (I can't build them, my latex refuses). Just open ``latext.ps`` in ai for example.

![image](https://user-images.githubusercontent.com/537321/102735997-bee8ee80-433b-11eb-9b16-4f8ee357711f.png)

PaulWessel commented:
I will see what my AI does but AI in general does not read PS well.

joa-quim commented:
Same happens with PDF

PaulWessel commented:
So ghostview and Preview works find while AI complains about fonts.  May have to look at PS level 1 font packages for this.  AI is a useless and overpriced tool.

PaulWessel commented:
This is what I was basing things on:

http://mirrors.ibiblio.org/CTAN/macros/latex/required/psnfss/psnfss2e.pdf

PaulWessel commented:
Some related information here the problem with AI: https://tex.stackexchange.com/questions/544617/use-standard-postscript-fonts-in-latex-helvetica-not-nimbussans

PaulWessel commented:
And this seems like the solution: Need to let AI know where all those fonts you have on your system live.

meghanrjones commented:
> I have added a new appendix and reference it from -B and pstext. However, I get this message
> 
> `/Users/pwessel/GMTdev/gmt-dev/dbuild/doc/rst/source/cookbook/gmt-latex.rst: WARNING: document isn't included in any toctree`

@PaulWessel, you can fix this by adding "cookbook/gmt-latex" to the toctree in doc/rst/source/cookbook.rst (starts at Line 35, postscript-fonts is at L50)

meghanrjones commented on doc/rst/source/text.rst:
```diff
@@ -73,6 +69,10 @@ To add a typeset figure caption for a 3-inch wide illustration, use
 
+To place a line containing a Latex equation, try::
+
+    echo 3 3 'Use @$\Delta g = 2\pi\rho Gh@$' | gmt text -R0/6/0/6 -JX15c -B -F+f32p+a30 -pdf map
```
```suggestion
    echo 3 3 'Use @[\Delta g = 2\pi\rho Gh@[' | gmt text -R0/6/0/6 -JX15c -B -F+f32p+a30 -pdf map
```
There's still a few places with the old syntax, not sure if that's just because the changes are in progress.

meghanrjones commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,38 @@
+==============================
+
+GMT supports the use of LaTeX equations embedded in text strings that are used
+with **-B** in titles, subtitles and axis labels, as well as for single lines
```
Is this statement correct? For me, LaTex equations work in titles, but not subtitles.

PaulWessel commented:
Yes, might have overdone that section.  WIll see, I think if possible we want this to work in subtitles.  Busy day today unfortunately.

meghanrjones commented:
> Yes, might have overdone that section. WIll see, I think if possible we want this to work in subtitles. Busy day today unfortunately.

Sounds good. If that doesn't work, it will just be necessary to change that sentence and would be helpful to add a sentence for -B specifying that **+s** is not possible if there is a LaTeX equation in **+t**.

meghanrjones commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -320,7 +320,8 @@ Append **+n** to have no frame and annotations at all [Default is controlled by
 **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-using the markers '@^' or '#break;'.
+using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title,
+enclose the expression with @$ markers (**Note**: Requires ``latex`` and ``dvips`` to be installed).
```
```suggestion
enclose the expression with @[ markers (**Note**: Requires ``latex`` and ``dvips`` to be installed).
```

meghanrjones commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -345,6 +346,8 @@ but you may also split this into two separate invocations for clarity, i.e.,
     force a horizontal label for *y*-axes (useful for very short labels).
     For Cartesian axes you may specify an alternate via **+s** which is used for
     right or upper axis axis label (with any **+l** label used for left and bottom axes).
+    To include LaTeX code as part of a label, enclose the expression with @$ markers
```
```suggestion
    To include LaTeX code as part of a label, enclose the expression with @[ markers
```

PaulWessel commented:
Thanks!

meghanrjones commented on src/pstext.c:
```diff
@@ -296,6 +296,7 @@ static int usage (struct GMTAPI_CTRL *API, int level) {
 	GMT_Message (API, GMT_TIME_NONE, "\t   @!<char1><char2> makes one composite character.\n");
 	GMT_Message (API, GMT_TIME_NONE, "\t   @. prints the degree symbol.\n");
 	GMT_Message (API, GMT_TIME_NONE, "\t   @@ prints the @ sign itself.\n");
+	GMT_Message (API, GMT_TIME_NONE, "\t   @$<LaTeX expression>@$ may be used (except for -M).\n");
```
```suggestion
	GMT_Message (API, GMT_TIME_NONE, "\t   @[<LaTeX expression>@[ may be used (except for -M).\n");
```

PaulWessel commented:
@joa-quim the issue with AI has to do with access to the fonts.  See the link about on how to use ghostscript fonts in AI.

meghanrjones commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,38 @@
+via package managers.  Your installation must have support for the packages *fontenc*
+and *inputenc*, as well as the required fonts *helvet, mathptmx, courier, symbol,
+avantgar, bookman, newcent, mathpazo, zapfchan* and *zapfding*.  If the conversion
+from LaTeX to EPS fails, please follow instructions to run the conversion script
```
I find the statement
>please follow instructions to run the conversion script manually' 

confusing because there is no information about where these instructions can be found. I think it would be better to say 'please run the conversion script manually' or insert a link to some instructions.

PaulWessel replied:
The instructions would be printed out when things fails, including the path to the temp directory holding all the files and logs etc.  Maybe "... follow instructions in the error message to run..." ?

seisman replied:
It would be good if you can write some technical details about how the GMT-LaTeX integration works, i.e., tex -> dvi -> eps -> psimage.

meghanrjones replied:
Yes, that change would make it clear

PaulWessel replied:
OK, once I finish another branch ... I will add an example of a typical tex file and the steps.

seisman commented on test/pstext/latext.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
Note that those tests will fail in our CI, because we don't have latex installed.

PaulWessel replied:
Can we flag that script to not run in the CI perhaps?

seisman replied:
I think we can check if `latex` and `dvips` are available. If not (for CI), then exclude these scripts from tests.

PaulWessel commented:
OK, I added the technical info but could need help with the formatting of the verbatim text, @seisman .

seisman commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,82 @@
+
+To help anyone debug their LaTeX installation, consider the case where you make a basemap
+that contains the title request -B+t"Use @[\\Delta g = 2\\pi\\rho Gh@[". This request ends
+up creating a temporary directory containing a small LaTex file called *gmt_eq.tex*:
+
+``\documentclass{article}``
+
+``\usepackage[T1]{fontenc} \usepackage[utf8]{inputenc}``
+
+``\usepackage{helvet}``
+
+``\begin{document}``
+
+  ``\thispagestyle{empty}``
+
+  ``\fontfamily{phv}\selectfont``
+
+  ``Use $\Delta g = 2\pi\rho Gh$``
+
+``\end{document}``
```
```suggestion
up creating a temporary directory containing a small LaTex file called *gmt_eq.tex*::

    \documentclass{article}
    \usepackage[T1]{fontenc} \usepackage[utf8]{inputenc}
    \usepackage{helvet}
    \begin{document}
      \thispagestyle{empty}
      \fontfamily{phv}\selectfont
      Use $\Delta g = 2\pi\rho Gh$
    \end{document}
```

seisman commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,82 @@
+
+Because :term:`FONT_TITLE` was set to Helvetica, the LaTeX file changes the default
+font to Helvetica as well (package *helvet*, code *phv*).  This file is then converted to
+a DVI file by the command
+
+``latex -interaction=nonstopmode gmt_eq.tex > /dev/null``
```
```suggestion
a DVI file by the command::

    latex -interaction=nonstopmode gmt_eq.tex > /dev/null
```

seisman commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,82 @@
+
+``latex -interaction=nonstopmode gmt_eq.tex > /dev/null``
+
+followed by the conversion to EPS via
+
+``dvips -q -E gmt_eq.dvi -o equation.eps``
+
```
```suggestion
followed by the conversion to EPS via::

    dvips -q -E gmt_eq.dvi -o equation.eps
```

PaulWessel commented:
Added the ability for subtitles to have LaTeX equations.  The only real restriction now is any title or subtitles with LaTeX must be single-line.  Added new test script, giving this:

![subtitle_latex](https://user-images.githubusercontent.com/26473567/103033061-acbaac00-4505-11eb-87a9-28aa341b1d71.png)

Perhaps this branch is now ripe for final review and merging? All tests pass (locally).

seisman commented:
Some warnings:
```
../../src/gmt_plot.c:275:42: warning: unused parameter 'GMT' [-Wunused-parameter]
bool gmt_text_is_latex (struct GMT_CTRL *GMT, const char *string) {
                                         ^
../../src/gmt_plot.c:284:59: warning: unused parameter 'GMT' [-Wunused-parameter]
GMT_LOCAL bool gmtplot_has_title_breaks (struct GMT_CTRL *GMT, const char *string) {
                                                          ^
../../src/gmt_plot.c:5907:123: warning: variable 'w' is uninitialized when used here [-Wuninitialized]
        GMT_Report (GMT->parent, GMT_MSG_DEBUG, "gmtplot_place_latex_eps: Conversion of LaTeX string gave dimensions %g x %g\n", w, h);
                                                                                                                                 ^
../../src/gmt_plot.c:5899:10: note: initialize the variable 'w' to silence this warning
        double w, h;
                ^
                 = 0.0
../../src/gmt_plot.c:5907:126: warning: variable 'h' is uninitialized when used here [-Wuninitialized]
        GMT_Report (GMT->parent, GMT_MSG_DEBUG, "gmtplot_place_latex_eps: Conversion of LaTeX string gave dimensions %g x %g\n", w, h);
                                                                                                                                    ^
../../src/gmt_plot.c:5899:13: note: initialize the variable 'h' to silence this warning
        double w, h;
                   ^
                    = 0.0
4 warnings generated.
```

seisman commented on doc/rst/source/pstext.rst:
```diff
@@ -51,21 +51,15 @@ Examples
-   ::
+To plot a text at the upper left corner of a 10 cm map::
 
     echo TopLeft | gmt pstext -R1/10/1/10 -JX10 -F+cTL -P > plot.ps
```
```suggestion
    echo TopLeft | gmt pstext -R1/10/1/10 -JX10 -B -F+cTL -P > plot.ps
```

meghanrjones replied:
Don't this and the next pstext example need ```-Baf``` to avoid warnings if gmt.history doesn't already exist?

seisman replied:
Oh yes, I forgot to delete the gmt.history file.

seisman commented on doc/rst/source/text.rst:
```diff
@@ -49,15 +49,11 @@ Examples
-   ::
+To plot a text at the upper left corner of a 10 cm map::
 
     echo TopLeft | gmt text -R1/10/1/10 -JX10 -F+cTL -pdf plot
```
```suggestion
    echo TopLeft | gmt text -R1/10/1/10 -JX10 -B -F+cTL -pdf plot
```

seisman commented on doc/rst/source/text_common.rst_:
```diff
@@ -23,7 +23,9 @@ Appendix :ref:`Chart-Octal-Codes-for-Chars` in the GMT Technical Reference and C
 (does not work for strings with sub/super scripts, symbols, or composite
-characters, except in paragraph mode (**-M**)).
+characters, except in paragraph mode (**-M**)).  Finally, you may typeset
+Latex expressions provided they are enclosed in @[ ... @[ or <math> ... </math>;
```
```suggestion
LaTeX expressions provided they are enclosed in @[ ... @[ or <math> ... </math>;
```

seisman commented on doc/rst/source/text_common.rst_:
```diff
@@ -137,6 +139,7 @@ Optional Arguments
     text dimensions (and **-C**) to build clip paths and turn clipping on.
     This clipping can then be turned off later with :doc:`clip` **-C**.
     To **not** plot the text but activate clipping, use **-G+n** instead.
+    **Note**: cannot be used with Latex expressions.
```
```suggestion
    **Note**: cannot be used with LaTeX expressions.
```

seisman commented on doc/rst/source/text_common.rst_:
```diff
@@ -162,7 +165,8 @@ Optional Arguments
     the justification set via **-F+j** applies to the box alignment
-    since the text justification is set by *parjust*.
+    since the text justification is set by *parjust*.  **Note**:
+    cannot be used with Latex expressions.
```
```suggestion
    cannot be used with LaTeX expressions.
```

seisman commented on doc/rst/source/text_common.rst_:
```diff
@@ -189,6 +193,7 @@ Optional Arguments
 **-W**\ *pen*
     Sets the pen used to draw a rectangle around the text string (see
     **-C**) [Default is width = default, color = black, style = solid].
+    **Note**: cannot be used with Latex expressions.
```
```suggestion
    **Note**: cannot be used with LaTeX expressions.
```

seisman commented on src/gmt_init.c:
```diff
@@ -6685,7 +6685,9 @@ void gmtlib_explain_options (struct GMT_CTRL *GMT, char *options) {
 			gmt_message (GMT, "\t     Append +t<title> to place a title over the map frame [no title]. Optionally also set +s<subtitle>.\n");
-			gmt_message (GMT, "\t     Note: Both <title> and <subtitle> can be set across multiple lines by using \"@^\" or \"#break;\" to mark breaks.\n");
+			gmt_message (GMT, "\t     Note: Both <title> and <subtitle> can be set across multiple lines by using \"@^\" or \"<break>\" to mark breaks.\n");
+			gmt_message (GMT, "\t     A single-line <title> may contain LaTeX code enclosed by @[ .... @[ (or alternatively <math> ... </math>).\n");
```
```suggestion
			gmt_message (GMT, "\t     A single-line <title> and <subtitle> may contain LaTeX code enclosed by @[ .... @[ (or alternatively <math> ... </math>).\n");
```

meghanrjones commented:
I recommend fixing the warning that 
```gmt/dbuild/doc/rst/source/cookbook/gmt-latex.rst: WARNING: document isn't included in any toctree```
by adding ```cookbook/gmt-latex``` to the toctree in ```doc/rst/source/cookbook.rst``` (the toctree starts at Line 35, gmt-latex could go right below postscript-fonts  at L50) before merging. GitHub only allows suggesting line changes to files that have been edited, which is why I cannot suggest this change directly through a PR review.

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -320,7 +320,8 @@ Append **+n** to have no frame and annotations at all [Default is controlled by
 will appear centered above the plot frame; optionally also add a *subtitle* via **+s**.
 **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-using the markers '@^' or '#break;'.
+using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title,
```
```suggestion
using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title or subtitle,
```

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -320,7 +320,8 @@ Append **+n** to have no frame and annotations at all [Default is controlled by
 **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-using the markers '@^' or '#break;'.
+using the markers '@^' or '#break;'.  To include LaTeX code as part of a single-line title,
+enclose the expression with @[ markers (**Note**: Requires ``latex`` and ``dvips`` to be installed).
```
```suggestion
enclose the expression with @[ markers (or alternatively <math> ... </math>).
(**Note**: Requires ``latex`` and ``dvips`` to be installed).
```

seisman commented on doc/rst/source/cookbook/options.rst:
```diff
@@ -345,6 +346,8 @@ but you may also split this into two separate invocations for clarity, i.e.,
     force a horizontal label for *y*-axes (useful for very short labels).
     For Cartesian axes you may specify an alternate via **+s** which is used for
     right or upper axis axis label (with any **+l** label used for left and bottom axes).
+    To include LaTeX code as part of a label, enclose the expression with @[ markers
```
```suggestion
    To include LaTeX code as part of a label, enclose the expression with @[ markers
    (or alternatively <math> ... </math>)
```

seisman commented on doc/rst/source/explain_-B_full.rst_:
```diff
@@ -30,8 +30,10 @@
     **Note**: Both *title* and *subtitle* may be set over multiple lines by breaking them up
-    using the markers '@^' or '#break;'.  The Frame setting is optional but
-    can be invoked once to override the above defaults.
+    using the markers '@^' or '<break>'.  To include LaTeX code as part of a single-line title,
```
```suggestion
    using the markers '@^' or '<break>'.  To include LaTeX code as part of a single-line title or subtitle,
```

seisman commented on doc/rst/source/cookbook/gmt-latex.rst:
```diff
@@ -0,0 +1,74 @@
+
+To help anyone debug their LaTeX installation, consider the case where you make a basemap
+that contains the title request -B+t"Use @[\\Delta g = 2\\pi\\rho Gh@[". This request ends
+up creating a temporary directory containing a small LaTex file called *gmt_eq.tex*::
```
```suggestion
up creating a temporary directory containing a small LaTeX file called *gmt_eq.tex*::
```

PaulWessel commented:
Also fixed that remaining `#break;` to` <break>`

seisman commented:
@PaulWessel Do you mind moving the 4 latex tests into a separate directory, e.g, `latex`? It would be easier to exclude them from the tests if latex or dvips are not found.

PaulWessel commented:
> @PaulWessel Do you mind moving the 4 latex tests into a separate directory, e.g, `latex`? It would be easier to exclude them from the tests if latex or dvips are not found.

Good idea, will do soon but in the middle of Xcode debug on another branch...  Or you can do it, create a latex test folder and move the ones from psbasemap and pstext I think.  Not sure about doc/scripts/GMT_latex.sh though as not a test.

seisman commented:
After installing the full mactex via homebrew `brew install --cask mactex-no-gui`, the LaTeX feature now works well for me.

There are still a few comments to address:
- https://github.com/GenericMappingTools/gmt/pull/4566#issuecomment-750464163
- https://github.com/GenericMappingTools/gmt/pull/4566#issuecomment-750459964

We may also need to add instructions to install LaTeX in the wiki.

PaulWessel commented:
> I recommend fixing the warning that
> `gmt/dbuild/doc/rst/source/cookbook/gmt-latex.rst: WARNING: document isn't included in any toctree`
> by adding `cookbook/gmt-latex` to the toctree in `doc/rst/source/cookbook.rst` (the toctree starts at Line 35, gmt-latex could go right below postscript-fonts at L50) before merging. GitHub only allows suggesting line changes to files that have been edited, which is why I cannot suggest this change directly through a PR review.

Thanks, now committed

PaulWessel commented:
> After installing the full mactex via homebrew `brew install --cask mactex-no-gui`, the LaTeX feature now works well for me.
> 
> There are still a few comments to address:
> 
> * [#4566 (comment)](https://github.com/GenericMappingTools/gmt/pull/4566#issuecomment-750464163)
> * [#4566 (comment)](https://github.com/GenericMappingTools/gmt/pull/4566#issuecomment-750459964)
> 
> We may also need to add instructions to install LaTeX in the wiki.

Warnings fixed.  instructions for wiki next but not in this repo.

PaulWessel commented:
Yippie!

seisman commented:
FYI, I created a wiki page for "how to install LaTeX" at https://github.com/GenericMappingTools/gmt/wiki/Install-LaTeX-for-GMT-LaTeX-Integration.
